### PR TITLE
Draft esm resolver specification

### PR DESCRIPTION
**Pull request by guybedford**
In line with the current plans for Phase 2 I've put together a draft resolver algorithm here that includes the new properties from the minimal kernel implementation.

The goal is, like the minimal kernel, to start with what we have agreed on, and to expand this over time as we can reach consensus.

This one brings up the discussions around the package.json main handling.

Would be nice to discuss this at the meeting today if there is time, although I understand it is late notice to read up on the details here.

### CODE REVIEW

**Comment by devsnek** [comment]
I think a better place for this would be doc/esm.md in our repo

**Comment by guybedford** [comment]
Rendered version here - https://github.com/nodejs/modules/blob/esm-resolver-spec/doc/esm-resolver-specification.md.

**Comment by guybedford** [comment]
@devsnek good point, I can redraft the PR there.

**Comment by jkrems** [comment]
I think having a more formal spec here that we can harden and then turn it into a more user-facing documentation makes sense. `doc/esm.md` is user documentation, not spec imo.

File: doc/esm-resolver-specification.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+1. _ESM_RESOLVE(@pkg/name, file:///path/to/project)_ -> _file:///path/to/node_modules/@pkg/name/index.mjs_ (if it exists)
+1. _ESM_RESOLVE(@pkg/name, file:///path/to/project)_ -> _file:///path/node_modules/@pkg/name/index.mjs_ (otherwise, if it exists)
+1. _ESM_RESOLVE(@pkg/name, file:///path/to/project)_ -> _file:///node_modules/@pkg/name/index.mjs_ (otherwise, if it exists)
+1. _ESM_RESOLVE(@pkg/name, file:///path/to/project)_ -> _Not Found_ (otherwise)
+
+### Resolver Algorithm
+
+The algorithm to resolve an ES module specifier is provided through _ESM_RESOLVE_:
+
+ESM_RESOLVE(specifier, parentURL)
```
Comment (jkrems) [inline doc/esm-resolver-specification.md]: Should we copy the beginning from the HTML spec even though the observable behavior *should* be the same?

> 1. Apply the URL parser to specifier. If the result is not failure, return the result.
> 2. If specifier does not start with the character U+002F SOLIDUS (/), the two-character sequence U+002E FULL STOP, U+002F SOLIDUS (./), or the three-character sequence U+002E FULL STOP, U+002E FULL STOP, U+002F SOLIDUS (../), return `<our custom logic>`.
> 3. Return the result of applying the URL parser to specifier with base URL as the base URL.

That ways it's a bit easier to tell where we deviate from what would happen in a browser.

**Comment by guybedford** [comment]
Per @devsnek's suggestion I've created a PR at https://github.com/nodejs/ecmascript-modules/pull/12/files. Since we are keeping these docs maintained on the minimal implementation I think that would make sense given that the specification for the CommonJS resolver _is user-facing documentation_.

**Comment by guybedford** [comment]
Closing this PR for now, let's continue discussion on the ecmascript-modules repo rather then.

**Pull request closed without merging**